pkg/kubectl/cmd/flags/record.go

```diff
@@ -0,0 +1,102 @@
+package flags
+
+import (
+	"github.com/spf13/cobra"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/kubectl"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+)
+
+// RecordFlags contains all flags associated with the "--record" operation
+type RecordFlags struct {
+	Record *bool
+
+	flagChanged bool
+	changeCause string
+}
+
+// ToRecorder returns a ChangeCause recorder if --record=false was not
+// explicitly given by the user
+func (f *RecordFlags) ToRecorder() (Recorder, error) {
+	shouldRecord := false
+	if f.Record != nil {
+		shouldRecord = *f.Record
+	}
+
+	// if flag was explicitly set to false by the user,
+	// do not record
+	if !shouldRecord && f.flagChanged {
+		return &NoopRecorder{}, nil
+	}
+
+	return &ChangeCauseRecorder{
+		shouldRecord: shouldRecord,
+		changeCause:  f.changeCause,
+	}, nil
+}
+
+func (f *RecordFlags) Complete(cmd *cobra.Command, changeCause string) error {
+	f.flagChanged = cmd.Flags().Changed("record")
+	f.changeCause = changeCause
+	return nil
+}
+
+func (f *RecordFlags) AddFlags(cmd *cobra.Command) {
+	if f.Record != nil {
+		cmd.Flags().BoolVar(f.Record, "record", *f.Record, "Record current kubectl command in the resource annotation. If set to false, do not record the command. If set to true, record the command. If not set, default to updating the existing annotation value only if one already exists.")
+	}
+}
+
+func NewRecordFlags() *RecordFlags {
+	record := false
+
+	return &RecordFlags{
+		Record: &record,
+	}
+}
+
+type Recorder interface {
+	Record(*resource.Info) error
+}
+
+type NoopRecorder struct{}
+
+func (r *NoopRecorder) Record(info *resource.Info) error {
+	return nil
+}
+
+// ChangeCauseRecorder annotates a "change-cause" to an input runtime object
+type ChangeCauseRecorder struct {
+	shouldRecord bool
+	changeCause  string
+}
+
+// Record annotates a "change-cause" to a given info if either "shouldRecord" is true,
+// or the resource info previously contained a "change-cause" annotation.
+func (r *ChangeCauseRecorder) Record(info *resource.Info) error {
```

Comment:
A problem I foresee with this signature is that the run command does not have a resource info to pass here. Maybe have the method receive a runtime.Object and a mapping instead? This is assuming that this method is the correct place to do the check below.